activerecord/lib/active_record/insert_all.rb

```diff
@@ -0,0 +1,185 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  class InsertAll
+    def initialize(model, inserts, **options)
+      raise ArgumentError, "inserts should be an array" unless inserts.is_a?(Array)
+      @model, @connection, @options = model, model.connection, options
+      @keys, @values_list = connection.prepare_keys_and_values_for_insert(inserts, model.table_name)
+    end
+
+    def execute
+      if values_list.empty?
+        ActiveRecord::Result.new([], [])
+      else
+        connection.exec_query to_sql, "Bulk Insert"
+      end
+    end
+
+    def to_sql
+      sql_builder.build_sql(
+        model,
+        keys,
+        values_list,
+        returning: returning,
+        on_duplicate: on_duplicate,
+        unique_by: unique_by,
+        updatable_columns: updatable_columns)
+    end
+
+    private
+      attr_reader :model, :connection, :options, :keys, :values_list
+
+      def sql_builder
+        connection.insert_all_sql_builder
+      end
+
+      def returning
+        returning = options.fetch(:returning) { connection.supports_insert_returning? && primary_keys }
+        returning = false if returning == []
```

Comment:
When is returning going to be an empty array?

Reply:
When `primary_keys` is an empty array (as on e.g. a join table).

—but I was also thinking that, from Rails' customers' POV, these two statements mean the same thing (and that from both inputs we _know_ what their expecting the output to be — so there's no harm in silently supporting the latter):

```ruby
Book.insert_all books, returning: false
Book.insert_all books, returning: []
```